### PR TITLE
[Follower] Dropping redundant inputs

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/4125

### Context

As part of cleanup of TODOs for consensus follower upgrade this PR updates how we process incoming batches.
There are two scenarios:
- (i) the blocks are all already known 
- (ii) there were some previously unknown blocks in the batch.

In scenario (i) we will drop the batch and resulting `certifiedBatch` will be empty.
In scenario (ii) we will process the batch as it is even if some blocks are already known.

### CODE REVIEW

**Review by jordanschalm — APPROVED** [review]
🎸

File: engine/common/follower/cache/cache.go [hunk 1]
```diff
@@ -151,7 +155,12 @@ func (c *Cache) AddBlocks(batch []*flow.Block) (certifiedBatch []*flow.Block, ce
 	//    (result stored in `batchContext.batchParent`)
 	//  * check whether last block in batch has a child already in the cache
 	//    (result stored in `batchContext.batchChild`)
+	//  * check if input is redundant, meaning that ALL blocks are already known
+	//    (result stored in `batchContext.redundant`)
```
Comment (AlexHentschel) [inline engine/common/follower/cache/cache.go]:
```suggestion
	//  * check if input is redundant (indicated by `batchContext.redundant`), i.e. ALL blocks
	//    are already known: then skip further processing
```

File: engine/common/follower/cache/cache.go [hunk 2]
```diff
@@ -305,6 +320,7 @@ func (c *Cache) cache(blockID flow.Identifier, block *flow.Block) (equivocation
 	if !added { // future proofing code: we allow an overflowing HeroCache to potentially eject the newly added element.
 		return
 	}
+	stored = true
```
Comment (AlexHentschel) [inline engine/common/follower/cache/cache.go:326]: would suggest to assign the return value of https://github.com/onflow/flow-go/blob/aa4d24b0b9d098219b67211b345795f43bb71b87/engine/common/follower/cache/cache.go#L319 directly to `stored`

[Lines 319-L323](https://github.com/onflow/flow-go/blob/aa4d24b0b9d098219b67211b345795f43bb71b87/engine/common/follower/cache/cache.go#L319-L323) would then be
```golang
	stored = c.backend.Add(blockID, block)
	if !stored { // future proofing code: we allow an overflowing HeroCache to potentially eject the newly added element.
		return
	}
```

**Review by AlexHentschel — APPROVED** [review]
clean, well-tested revision. Very nice. 

I took the liberty to polish and extend a few code comments (mostly in areas that are not touched by your code) 👉 https://github.com/onflow/flow-go/pull/4284/commits/ad4b98948195743df71fa4d83458d91f44a0df7f

**Comment by durkmurder** [comment]
bors merge

**Comment by durkmurder** [comment]
bors merge

**Pull request merged**